### PR TITLE
Add new filamat_lite target to filamat

### PR DESCRIPTION
**Pull request by bejado**
This is the first pass at reducing the size of the `filamat` library. This adds a new target
`filamat_lite`, which has fewer dependencies and a smaller binary size. Its restrictions:

1. Only supports OpenGL backend
1. No material optimizatione
1. No compile-time static analysis of shader code
1. Stricter requirements for setting properties in the `material` function (see
   `GLSLToolsLite.h`)

Size reduction:

| Architecture | filamat | filamat_lite | % reduction |
| -- | -- | -- | -- |
| arm64-v8a/libfilamat-jni.so | 5107 KiB | 842 KiB | 16.5% |
| armeabi-v7a/libfilamat-jni.so | 3466 KiB | 462 KiB | 13.3% |
| x86/libfilamat-jni.so | 5358 KiB | 874 KiB | 16.3% |
| x86_64/libfilamat-jni.so | 5547 KiB | 879 KiB | 15.8% |

How including filamat_lite in the filament-jni library affects binary size:

| Architecture | Size without filamat_lite | Size increase with filamat_lite | % Size increase |
| -- | -- | -- | -- |
| arm64-v8a/libfilament-jni.so | 742 KiB | +580 KiB | 78.1% |
| armeabi-v7a/libfilament-jni.so | 522 KiB | +336 KiB | 64.4% |
| x86/libfilament-jni.so | 810 KiB | +624 KiB | 77.1% |
| x86_64/libfilament-jni.so | 787 KiB | +612 KiB | 77.8% |

### CODE REVIEW

File: android/filamat-android/CMakeLists.txt [hunk 1]
```diff
@@ -2,9 +2,14 @@ cmake_minimum_required(VERSION 3.4.1)
 
 set(FILAMENT_DIR ${FILAMENT_DIST_DIR})
 
-add_library(filamat STATIC IMPORTED)
-set_target_properties(filamat PROPERTIES IMPORTED_LOCATION
-        ${FILAMENT_DIR}/lib/${ANDROID_ABI}/libfilamat.a)
+set(FILAMAT_VERSION "filamat")
```
Comment (prideout) [inline android/filamat-android/CMakeLists.txt]: It's not really a version, it's more like a flavour...maybe simply call this `FILAMAT_NAME`?  Not a big deal.

Comment (bejado) [reply android/filamat-android/CMakeLists.txt]: Good suggestion!

File: android/filamat-android/build.gradle [hunk 2]
```diff
@@ -65,6 +65,23 @@ android {
         }
     }
 
+    flavorDimensions "functionality"
```
Comment (romainguy) [inline android/filamat-android/build.gradle]: Nice :)

File: libs/filamat/README.md [hunk 3]
```diff
@@ -170,3 +173,48 @@ material with Filament, pass the material package's data into a Filament Materia
...
+// The MaterialInputs variable must be named material.
+void material(inout MaterialInputs material) {
+    prepareMaterial(material);
+
+    // Good.
+    material.roughness = materialParams.roughness;
+    material.baseColor.rgb = vec3(1.0, 0.0, 1.0);
+
+    // Incorrect! While this line is technically correct GLSL code, the simple parser inside
+    // filamat_lite will not pick up the set to clearCoat due to the additional whitespace.
```
Comment (romainguy) [inline libs/filamat/README.md]: Sounds like something we could easily support, although probably not worth it

File: libs/filamat/README.md [hunk 4]
```diff
@@ -170,3 +173,48 @@ material with Filament, pass the material package's data into a Filament Materia
...
+    material . clearCoat  = 1.0;
+
+    // Warning! The code is commented-out, but the parser still consideres the anisotropy
```
Comment (romainguy) [inline libs/filamat/README.md]: Should we handle comments?

File: libs/filamat/src/sca/GLSLToolsLite.cpp [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                    MaterialBuilder::PropertyList& properties) const noexcept {
+    const std::string shaderCode(material.c_str());
+
+    size_t start = 0, end = 0;
+
+    const auto p = Enums::map<Property>();
+
+    // Find all occurrences of "material.someProperty" in the shader string.
+    size_t loc;
+    while ((loc = shaderCode.find("material.", start)) != std::string::npos) {
```
Comment (romainguy) [inline libs/filamat/src/sca/GLSLToolsLite.cpp]: We could do a first pass to just remove white spaces and comments

File: libs/filamat/README.md (hunk 3 above)
Comment (bejado) [reply libs/filamat/README.md]: Yeah, I originally had a `regex` which would handle this case, but `std::regex` pulls in ~40KiB or so.. I decided it wasn't worth it as it's probably an edge case

File: libs/filamat/README.md (hunk 4 above)
Comment (bejado) [reply libs/filamat/README.md]: Yeah, it's probably not too hard to do a pass to remove comments.

File: libs/filamat/src/sca/GLSLToolsLite.cpp (hunk 5 above)
Comment (bejado) [reply libs/filamat/src/sca/GLSLToolsLite.cpp]: Good idea- although I guess the only white spaces we care about are between the "material" token and any property sets on it. Maybe I can modify the search algorithm to allow whitespace.

**Comment by rawnsley** [comment]
What are the implications of having no material optimization? Basically I'm trying to work out if there will be a render performance hit from using filamat_list vs filamat.

**Comment by romainguy** [comment]
It shouldn't affect runtime performance. Binary materials might be bigger and shader compilation might be a bit slower on device.

**Comment by rawnsley** [comment]
Frustratingly shaders compiled without any optimization crash on Mali GPU (matc or either filamat library). I'll have a poke around and try and work out why.

**Comment by romainguy** [comment]
Any particular variant of Mali?

**Comment by rawnsley** [comment]
It's my old T760 on driver version 11. I've whittled it down to this vertex shader:
```
#version 300 es

invariant gl_Position;

struct MaterialVertexInputs {
    vec3 worldNormal;
    vec4 worldPosition;
};


void main() {
    MaterialVertexInputs material;
    gl_Position = material.worldPosition;
}
```
It causes a segmentation fault in the [offline compiler](https://developer.arm.com/tools-and-software/graphics-and-gaming/graphics-development-tools/mali-offline-compiler/downloads) as well. I've just got to try and work out what to do about it.

**Comment by romainguy** [comment]
Thank you! would you mind filing an issue? It could be an alignment problem. Could you try to swap the order of the fields in the struct?

**Comment by rawnsley** [comment]
Nice idea, but unfortunately it didn't change anything. I've raised an issue as suggested.

**Pull request merged**